### PR TITLE
Add vir::simd_permute following P2664

### PR DESCRIPTION
**Pull request by mattkretz**
ChangeLog:

	* README.md: Document simd_permute.
	* vir/simd.h (load & stores): Make constexpr.
	* vir/simd_concepts.h (any_simd_or_mask): New.
	* vir/simd_permute.h: New file.
	* vir/test.cpp (make_simd): New. Add simd_permute tests.

### CODE REVIEW

File: vir/simd.h [hunk 1]
```diff
@@ -513,24 +513,26 @@ namespace vir::stdx
 
       // load constructor
       template <typename Flags>
+        constexpr
```
Comment (ChristianTackeGSI) [inline vir/simd.h:516]: I would suggest to put this part into its own commit.

File: vir/test.cpp [hunk 2]
```diff
@@ -155,6 +241,7 @@ static_assert(std::same_as<vir::as_tuple_t<Line>,
 static_assert(std::same_as<vir::simdize<Line>,
 			   vir::simd_tuple<Line, V<float>::size()>>);
 
+#if VIR_HAVE_SIMD_IOTA
```
Comment (ChristianTackeGSI) [inline vir/test.cpp:244]: This looks unrelated as well?

File: README.md [hunk 3]
```diff
@@ -116,6 +116,62 @@ Note that `vir::cvt` also works for `simd_mask` and non-`simd` types. Thus,
 (i.e. well-formed for `T` and `simd<T>`).
 
 
+### Permutations ([paper](https://wg21.link/P2664):
+
+*Requires Concepts (C++20).*
+
+```c++
+#include <vir/simd_perute.h>
```
Comment (ChristianTackeGSI) [inline README.md:124]: Typo?

```suggestion
#include <vir/simd_permute.h>
```

**Review by ChristianTackeGSI — COMMENTED** [review]
The main part of this PR will take much longer for me to look over. But here are some very quick first comments.

File: vir/test.cpp (hunk 2 above)
Comment (mattkretz) [reply vir/test.cpp:244]: Yes. A drive-by clean up that I don't believe warrants its own commit.

File: vir/simd.h (hunk 1 above)
Comment (mattkretz) [reply vir/simd.h:516]: Yea..ah okay. :) This is a behavior enhancement...

File: vir/simd_permute.h [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	}
+      };
+
+    template <int Offset>
+      inline constexpr Shift<Offset> shift {};
+  }
+
+  template <std::size_t N = 0, vir::any_simd_or_mask V,
+	    detail::index_permutation_function<V::size()> F>
+    VIR_ALWAYS_INLINE constexpr stdx::resize_simd_t<N == 0 ? V::size() : N, V>
```
Comment (ChristianTackeGSI) [inline vir/simd_permute.h:140]: Why "always" inline? Is there a likely case where the compiler would not inline this at all? Or do you want to avoid symbols in the final object that might depend on compiler flags and conflict with other symbols of the same name?

Anyway: Maybe explain this somewhere?

File: vir/test.cpp [hunk 5]
```diff
@@ -53,6 +63,82 @@ namespace
...
+  all_of(vir::simd_permute(make_simd(0, 1, 2, 3), vir::simd_permutations::duplicate_odd)
+	   == make_simd(1, 1, 3, 3)));
+
+static_assert(
+  all_of(vir::simd_permute(make_simd(0, 1, 2, 3), vir::simd_permutations::swap_neighbors<>)
+	   == make_simd(1, 0, 3, 2)));
+
+static_assert(
+  all_of(vir::simd_permute(make_simd(0, 1, 2, 3), vir::simd_permutations::swap_neighbors<2>)
+	   == make_simd(2, 3, 0, 1)));
```
Comment (ChristianTackeGSI) [inline vir/test.cpp:81]: At first this was surprising to me. (I had expected something else. I can't say exactly what though.)

Anyway: Are there simd-types (maybe with T=char) that have 6 elements, then maybe include an example with `vir::simd_permutations::swap_neighbors<3>` as well?

**Review by ChristianTackeGSI — COMMENTED** [review]
Looks good to me. I have some more comments.

File: vir/simd_permute.h (hunk 4 above)
Comment (mattkretz) [reply vir/simd_permute.h:140]: debug builds. Consider `simd` to be the missing fundamental type. You don't get function calls for builtin operations on `int` either. There's also constant propagation to consider here. A combination of operations involving one (or more) permute(s) might be optimized into something very different by the compiler if it considers everything together. Also I've been bitten recently: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108030. If the compiler gets it wrong you can pay in factors of performance, not percentages.

File: vir/test.cpp (hunk 5 above)
Comment (mattkretz) [reply vir/test.cpp:81]: good point swapping 2 on 4 elements is still ambiguous. will do

File: vir/simd_permute.h [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+    template <int Offset>
+      struct Shift
+      {
+	consteval int
+	operator()(int i, int size) const
+	{
+	  const int j = i + Offset;
+	  if constexpr (Offset >= 0)
+	    return j >= size ? simd_permute_zero : j;
```
Comment (ChristianTackeGSI) [inline vir/simd_permute.h:128]:
```cpp
// in Rotate::operator():
return (i + Offset) % size();
// in Shift::operator():
return j >= size ? simd_permute_zero : j;
```

Can you please explain, why `size` needs a `()` in one case and not in the other?

**Review by ChristianTackeGSI — COMMENTED** [review]
last bits…

File: vir/simd_permute.h (hunk 6 above)
Comment (mattkretz) [reply vir/simd_permute.h:128]: `size` is of type `integral_constant<int, Value>`. It is therefore convertible to `int` and has a call operator that returns `int`. Both are equivalent.
... unless it's an operand to an overloaded operator that deduces the operand as `integral_constant`, in which case it can use its value in a constant expression.

My use of `integral_constant` is not consistent yet. I'm still figuring out what I like best.

Comment (mattkretz) [reply vir/simd_permute.h:128]: Ah, actually in line 128 it's an `int size`, so not callable. Just a plain `int`.
